larpandoracontent/LArHelpers/LArHitWidthHelper.cc

```diff
@@ -0,0 +1,411 @@
+/**
+ *  @file   larpandoracontent/LArHelpers/LArHitWidthHelper.cc
+ *
+ *  @brief  Implementation of the lar hit width helper class.
+ *
+ *  $Log: $
+ */
+
+#include "larpandoracontent/LArHelpers/LArHitWidthHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+LArHitWidthHelper::ConstituentHit::ConstituentHit(const CartesianVector &positionVector, const float hitWidth, const Cluster *const parentClusterAddress) : m_positionVector(positionVector), m_hitWidth(hitWidth), m_parentClusterAddress(parentClusterAddress)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const Cluster* LArHitWidthHelper::ConstituentHit::GetParentClusterAddress() const
+{
+    if (!m_parentClusterAddress)
+        throw StatusCodeException(STATUS_CODE_NOT_INITIALIZED);
+
+    return m_parentClusterAddress;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool LArHitWidthHelper::ConstituentHit::SortByDistanceToPoint::operator() (const ConstituentHit &lhs, const ConstituentHit &rhs)
+{
+    CartesianVector lhsPosition(lhs.GetPositionVector());
+    CartesianVector rhsPosition(rhs.GetPositionVector());
+
+    return (m_referencePoint.GetDistanceSquared(lhsPosition) < m_referencePoint.GetDistanceSquared(rhsPosition));   
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+    
+LArHitWidthHelper::ClusterParameters::ClusterParameters(const Cluster *const pCluster, const float maxConstituentHitWidth, const bool isUniformHits, const float hitWidthScalingFactor):
+    m_pCluster(pCluster),
+    m_numCaloHits(pCluster->GetNCaloHits()), 
+    m_constituentHitVector(isUniformHits ? LArHitWidthHelper::GetUniformConstituentHits(pCluster, maxConstituentHitWidth, hitWidthScalingFactor) : LArHitWidthHelper::GetConstituentHits(pCluster, maxConstituentHitWidth, hitWidthScalingFactor)),
+    m_totalWeight(LArHitWidthHelper::GetTotalClusterWeight(m_constituentHitVector)),
+    m_lowerXExtrema(LArHitWidthHelper::GetExtremalCoordinatesLowerX(m_constituentHitVector)), 
+    m_higherXExtrema(LArHitWidthHelper::GetExtremalCoordinatesHigherX(m_constituentHitVector))
```

Comment:
Okay, so there is a function LArClusterHelper::GetExtremalCoordinates which finds the extreme inner and outer coordinates (inner is the lower z and outer is higher z). I had copied this function and instead set the variables by their x values (which is handled in a few lines at the end). I do this because my alg walks along the x direction rather than the z direction. So now, i have replaced the copy with a function that calls the LArClusterHelper function and then assigns the extremal coordinates appropriately, according to X. I'll think about other reductions though.